Blocks/MMQ/mmpq.h

```diff
@@ -70,83 +70,39 @@ class MMPQ {
   }
 
   // Adds a message to the buffer. Supports both copy and move semantics. THREAD SAFE.
-
-  // NOTE(dkorolev): `std::enable_if_t<std::is_same<message_t, current::decay<T>>::value, idxts_t>` can't convert
-  // `const char*` into an `std::string`, which is essential, as, unlike MMQ, MMPQ is not an `ss::EntryPublisher<>`.
-  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock, class T>
-  idxts_t Publish(T&& message) {
-    return DoPublish<MLS>(std::move(message), current::time::DefaultTimeArgument());
-  }
-
-  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock, class T>
-  idxts_t Publish(T&& message, const std::chrono::microseconds us) {
-    return DoPublish<MLS>(std::move(message), us);
-  }
-
-  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock, class T>
-  idxts_t PublishIntoTheFuture(T&& message) {
-    return DoPublishIntoTheFuture<MLS>(std::move(message), current::time::DefaultTimeArgument());
-  }
-
-  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock, class T>
-  idxts_t PublishIntoTheFuture(T&& message, std::chrono::microseconds us) {
-    return DoPublishIntoTheFuture<MLS>(std::move(message), us);
-  }
-
-  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock>
-  void UpdateHead(const std::chrono::microseconds us) {
-    DoUpdateHead<MLS>(us);
-  }
-
-  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock>
-  void UpdateHead() {
-    DoUpdateHead<MLS>(current::time::DefaultTimeArgument());
-  }
-
- private:
-  MMPQ(const MMPQ&) = delete;
-  MMPQ(MMPQ&&) = delete;
-  void operator=(const MMPQ&) = delete;
-  void operator=(MMPQ&&) = delete;
-
-  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock, class T, typename US>
-  idxts_t DoPublish(T&& message, const US us) {
+  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock,
+            class E,
+            typename TIMESTAMP>
+  idxts_t PublisherPublishImpl(E&& entry, TIMESTAMP&& timestamp) {
     locks::SmartMutexLockGuard<MLS> lock(mutex_);
-    const auto timestamp = current::time::GetTimestampFromLockedSection(us);
-    if (!(timestamp > last_idx_ts_.us)) {
-      CURRENT_THROW(ss::InconsistentTimestampException(last_idx_ts_.us + std::chrono::microseconds(1), timestamp));
-    }
+    const auto us = current::time::GetTimestampFromLockedSection(std::forward<TIMESTAMP>(timestamp));
```

Comment:
Good point. Renames and removed `forward`-s and extended references, in c977628.